README
Wikit is a "wiki wiki web" implementation, built with Tcl and MetaKit.
See http://www.equi4.com/starkit/wikit.html for details and news.

Wikit was written by Jean-Claude Wippler <jcw@equi4.com> and drives the
"Tcler's Wiki", an active collaborative resource <http://wiki.tcl.tk/>.

With many thanks to all contributors and authors of code used in Wikit:

      Steve Landers - improved markup for documentation, local mode
          Don Libes - cgi package
        Kevin Kenny - autoscroll package
    Andreas Kupries - documentation
    Colin McCormack - code refactoring
    Jos Decoster    - improved markup - incl. tables, backrefs, etc
        <your name> - <tell me>

Hold the author harmless and any lawful use is permitted.

-jcw, April 2007

IMPORTANT - Since 2007 the development of lib/wikit now occurs via
the Wub project at https://wikitcl.googlecode.com/svn/trunk/wikit

Wikit can be run in one of five modes
	- as a library via Wub - as in The Tclers Wiki (http://wiki.tcl.tk)
	- as a self-contained program via wikit's built-in web server 
	- as a CGI script run from Apache or TclHttpd or AOLserver
	- as a stand-alone program, with the rendering done by Tk
	- as an embedded portable help system for Starkits

The core wikit library code (lib/wikit) is now maintained via the
Wub/wikitcl project - i.e. being the code that runs the Tclers Wiki
(http://wiki.tcl.tk)

The remainder of this repository contains the code for running wikit
in the other modes, and as a Starkit - i.e. the dependent packages
(in ../lib) and the documentation (../doc).


ChangeLog
2011-08-18  jcw
   * The Subversion repository at svn://svn.equi4.com/wikit is now
     obsolete. GitHub has the new official repository for Wikit.
2007-04-10  jcw
   * added my name as original author to all relevant source files
2006-12-06  stevel
   * Changelog: improved markup for documentation
2006-12-06
   * Changelog: re-instated, cvstrac is not used after all
2006-01-12  jcw
   * gbutton/pkgIndex.tcl: fix for ticket #2, version set to 0.3
   * wikit/utils.tcl: make sure files are written in UTF-8 format
2005-11-03  jcw
   * *: tweaks to bring the system up to date (from Andreas Kupries)
2005-03-02  jcw
   * wikithttpd.tcl: typo in error message: "[h]ttpreturn", thx A. Groth
2005-02-01  jcw
   * web.tcl: avoid DoS in -httpd via GET /1000000000000000000000000000
2005-01-07  jcw
   * db.tcl: fix of bts#107, it failed with "-update"
   * image.tcl, db.tcl: added autoproxy support, see bts#108
     this also adds autoproxy, base64, and uri - taken from tcllib 1.7
2004-09-22  jcw
   * web.tcl: support WIKIT_WRU_REQ registered edits, thx Daniel Steffen
   also some tweaks w.r.t. cache flushing and better save message
2004-09-09  jcw
   * web.tcl: fix for incorrect WIKIT_BASE links, by Daniel Steffen
   * {cache,image,lock,search,utils}.tcl: remove trailing emacs comments
2004-08-20  jcw
   * cache.tcl: fix bug#97, prevented adding new page in httpd mode
2004-06-08  jcw
   * lots: refactored source code, changes by Colin McCormack
2004-05-31  jcw
   * format.tcl: change "set info [eval $ip [list $text]]" for Colin M.
2004-05-29  jcw
   * web.tcl: added new WIKIT_REVS var to add revision link to each page
2004-05-25  jcw
   * web.tcl: strip "Q=" prefix from queries so a form can be used
2004-05-19  jcw
   * web.tcl: changes to support usernames via cookies (not set here, but
     assumed to have been seen through other means, i.e. wru.tcl script).
     Relies on a new WIKIT_WRU env var to point to special decode script.
2004-03-25  jcw
   * web.tcl, wikidoc.tkd: minor tweaks for the March 4th changes, by PS.
2004-03-04  jcw
   * wikit.tcl, utils.tcl: changed edit and referenced links to have an
   edit/ cq reference/ URL prefix, so spiders can be told not to traverse.
   Idea by Donal Fellows, code changes by Pascal Scheffers - thanks!
2004-01-10  jcw
   * web.tcl: removed obsolete chdir (no ill effects in normal operation)
2003-11-11  jcw
   * wikithttpd.tcl: additional length tweak by TL to deal with MSIE
2003-11-10  jcw
   * wikithttpd.tcl: avoid Content-Length case sensitivity (thx T Lemler).
2003-10-29  jcw
   * start.tcl, wikithttpd.tcl, web.tcl: added changes to support r/w mode
     in the embedded http server mode previously added by Pascal Scheffer
     Many thanks to Michael Heca for this great enhancement.
2003-10-14  jcw
   * web.tcl: fixed a silly cache flush bug, thx AKG
2003-09-26  jcw
   * gui.tcl: start local browser for external links when on Windows, and
   add a status bar current link target (thx Bob Techentin, BTS#69+70)
2003-09-03  jcw
   * web.tcl: added AKG's small web tweaks, see http://mini.net/tcl/19
   * utils.tcl: cleanup the archive view and drop the format prop on open
2003-09-02  jcw
   * utils.tcl: fix error when the page name is a big number (thx John T)
2003-08-20  jcw
   * web.tcl: added css style for footer use (thx Daniel Steffen)
2003-06-20  jcw
   * format.tcl: get rid of extraneous spaces when creating page titles
2003-06-05  jcw
   * gui.tcl: fix to allow use of wheelmouse (thx Albrecht Mucha), bts#55
   * web.tcl: added WIKIT_EMAIL env var for error reporting (LV), bts#13
   * start.tcl,wikithttpd.tcl,web.tcl: encoding fixes, switch everything
     to utf-8, CGI and embedded mode (thx, Benjamin Riefenstahl), bts#53
2003-06-03  jcw
   * web.tcl: fix so back reference link in cached pages stay valid longer
2003-05-01  jcw
   * start.tcl: blech, the new "nolock" logic was wrong, fixed now
2003-04-30  jcw
   * web.tcl: forgot to do title quoting on unreferenced page, see Feb 12
2003-03-18  jcw
   * start.tcl: added "-nolock" cmdline option to prevent use of lockfile
2003-03-14  jcw
   * start.tcl, web.tcl: make wikit -httpd work on Win32 (thx to Steve L)
   * utils.tcl: graburl -binary 1, to avoid crlf with -update on windows
   * wikithttpd.tcl: added unset ::_cgi_uservar fix
2003-03-03  jcw
   * utils.tcl: fixed locking bug (thx Rohan P)
2003-02-27  jcw
   * gui.tcl: small tweaks to fix edit mode (by Tom K)
2003-02-21  jcw
   * start.tcl: fix -httpd handling, and catch unkmown options
   * web.tcl: change date and menu sizes (thx Reinhard M, suggestion#24)
2003-02-19  jcw
   * gui.tcl, modify,tcl, utils.tcl, web.tcl: drop special "format" flag
   * modify,tcl, utils.tcl: drop archive view, use the pages view to
   generate recent changes page from (multiple edits no longer listed)
   * gui.tcl: added "-bg black" changes (thx Bryan S, bug#22)
2003-02-12  jcw
   * utils.tcl, web.tcl: fixed two cases of quoting in page title names
2003-02-11  jcw
   * format.tcl: adjusted regex to pass "&#NNN;" to enable arabic, etc.
2003-02-06  jcw
   * support AOLserver: fix by Pascal Scheffers from http://mini.net/tcl/4416
2003-01-29  jcw
   * gui.tcl: tweaks to get edit mode wrapping right again (by Tom K)
2003-01-23  jcw
   * format.tcl, gui.tcl: changes contributed by Tom Krehbiel (thx!), to
   improve local Tk mode and make it much more consistent with CGI mode
2003-01-22  jcw
   * start.tcl, web.tcl, httpd/: new code to support running an internal
   httpd server (cli "-httpd port") - courtesy of Pascal Scheffers (thx!)
2002-12-08  jcw
   * start.tcl: added quick-intercept to return raw NNN.txt page requests
   * start.tcl: get rid of pages 10..up when initing a new wiki
2002-11-23  jcw
   * format.tcl: try loading new jpeg/png img format handlers (new tkimg)
   * modify.tcl: changed content sens bug in regsub (in ReplacLink proc)
2002-11-16  jcw
   * gui.tcl: changed "darkgrey" to "#404040" to port better (thanks Tom K)
2002-11-15  jcw

lib/app-wikit/start.tcl
# start.tcl -- Wikit startup code, locks and chooses between CGI and Tk mode
# originally written by Jean-Claude Wippler, 2000..2007 - may be used freely

package provide app-wikit 0.1
if {[catch { package require Mk4tcl }]} {
  package require mklite
  mklite::emulateMk4tcl
}

set roflag [lsearch -exact $argv "-readonly"]
if {$roflag >= 0} {
  set argv [lreplace $argv $roflag $roflag]
  incr argc -1
}

set uselock 1
set nolock [lsearch -exact $argv "-nolock"]
if {$nolock >= 0} {
  set argv [lreplace $argv $nolock $nolock]
  incr argc -1
  set uselock 0
}
unset nolock

set pos [lsearch -exact $argv "-images"]
if {$pos >= 0} {
  set imflag [lindex $argv [incr pos]]
  set argv [lreplace $argv [expr {$pos-1}] $pos]
  incr argc -2
}
unset pos

set pos [lsearch -exact $argv "-update"]
if {$pos >= 0} {
  set upflag [lindex $argv [incr pos]]
  set argv [lreplace $argv [expr {$pos-1}] $pos]
  incr argc -2
}
unset pos

# the wikit httpd server
set wikit_httpd ""
set pos [lsearch -exact $argv "-httpd"]
if {$pos >= 0} {
  set wikit_httpd [lindex $argv [incr pos]]
  set argv [lreplace $argv [expr {$pos-1}] $pos]
  incr argc -2
  # on Windows we need to display a console (if we can - we might be running
  # under tclkitsh)
  if {$::tcl_platform(platform) eq "windows"} {
    if {![catch {console show}]} {
      # probably tclkit - create a simple button to stop the server
      proc console_hide {} {
        console hide
        .c configure -text "Show wikit httpd server console" -command console_show
      }
      proc console_show {} {
        .c configure -text "Hide wikit httpd server console" -command console_hide
        console show
      }
      button .c
      button .e -text "Exit wikit httpd server" -command exit
      pack .c .e -fill x 
      wm title . "Wikit httpd server"
      console title "Wikit httpd server console"
      console_show
      after idle raise .
    }
  }
}
unset pos

# avoid things like "-help" from becoming a file name
if {[string index $argv 0] eq "-"} {
  error "unknown option: $argv"
}

if {$argc} {
  set wikidb [lindex $argv 0]
} else {
  # this should be derived from the basename of the executable that
  # invokes wikit
  set wikidb wikit.tkd
}

# 2002-12-08: quick intercept of NNN.txt page requests
catch {
  if {$env(REQUEST_METHOD) eq "GET" &&
      [regexp {/(\d+)\.txt$} $env(REQUEST_URI) - page]} {
    mk::file open xdb $wikidb -readonly
    puts "Content-type: text/plain\n"
    puts "Title:\t[mk::get xdb.pages!$page name]"
    puts "Date:\t[clock format [mk::get xdb.pages!$page date] \
    				-gmt 1 -format {%e %b %Y %H:%M:%S GMT}]"
    puts "Site:\t[mk::get xdb.pages!$page who]"
    puts ""
    puts -nonewline [mk::get xdb.pages!$page page]
    exit
  }
}

package require Wikit::Utils

# 2002-06-17: read everything before locking down the database
# this way, if a transfer takes long, it won't hold up other CGI's
# this is essential to avoid hanging on a "Moz 0.9.9 > 8 Kb edit" bug

if {[info exists ::env(SCRIPT_NAME)]} {
  fconfigure stdout -encoding utf-8
  fconfigure stdin -encoding utf-8
  encoding system utf-8
  package require cgi
  cgi_input
}

set appname [file rootname [file tail $wikidb]]

package require Wikit::Lock
if {$uselock && ![Wikit::AcquireLock $appname.lock]} {
  puts stderr "Can't lock: $appname.lock" ;# the laziest way out
  exit 1
}

# make sure the lock is always released
if {$uselock} {
  rename exit Wikit::exit
  proc exit {args} {
    global appname
    file delete $appname.lock
    catch {eval Wikit::exit $args}
    return
  }
}

if {[catch {
  package require Wikit::Format
  namespace import Wikit::Format::*

  package require Wikit::Db

  Wikit::WikiDatabase $wikidb

  if {[mk::view size wdb.pages] == 0} {
    # copy first 10 pages of the default datafile 
    set fd [open [file join $starkit::topdir doc wikidoc.tkd]]
    mk::file load wdb $fd
    close $fd
    mk::view size wdb.pages 10
    mk::view size wdb.archive 0
    Wikit::FixPageRefs
  }

  if {[info exists upflag]} {
    Wikit::DoSync $upflag
  }

  package require Wikit::Cache
  Wikit::BuildTitleCache

  if {[info exists ::env(SCRIPT_NAME)]} {
    package require Web
    eval [mk::get wdb.pages!9 page]
    if {[info exists ::env(WIKIT_ADMIN)]} {
      set ProtectedPages {}
    }
    Wikit::ProcessCGI
  } elseif { $wikit_httpd ne "" } { 
      encoding system utf-8
      package require cgi      
      package require Web
      package require wikithttpd

      #set ::httpd::debug 1
      eval [mk::get wdb.pages!9 page]
      if {[info exists ::env(WIKIT_ADMIN)]} {
	  set ProtectedPages {}
      }
      ::httpd::startServer localhost $wikit_httpd
      vwait forever
  } elseif {[info exists imflag]} {
    package require Wikit::Image
    Wikit::LocalImages $imflag
  } elseif {[catch {package require Wikit::Gui} msg]} {
    if {$uselock} { Wikit::ReleaseLock $appname.lock }
    exit 1
  } else {
    Wikit::LocalInterface 
  }
} err]} {
  puts stderr "error: $errorInfo"
}

if {$uselock} { Wikit::ReleaseLock $appname.lock }
